### PR TITLE
OGM-412 Align the use of house keeping method calls between ORMs EntityM...

### PR DESCRIPTION
**Pull request by hferentschik**
...anagerImpl/SessionImpl with the respective OgmEntityManager/OgmSessionImpl

### CODE REVIEW

File: core/src/main/java/org/hibernate/ogm/jpa/impl/OgmEntityManagerFactory.java [hunk 1]
```diff
@@ -39,22 +41,26 @@ public OgmEntityManagerFactory(EntityManagerFactory hibernateEmf) {
 
 	@Override
 	public EntityManager createEntityManager() {
-		return new OgmEntityManager( this, hibernateEmf.createEntityManager() );
+		return new OgmEntityManager( this, hibernateEmf.createEntityManager().unwrap( EntityManagerImpl.class ) );
```
Comment (Sanne) [inline core/src/main/java/org/hibernate/ogm/jpa/impl/OgmEntityManagerFactory.java]: There is no handling for the specific implementation _EntityManagerImpl.class_ in the unwrap() implementation of our EntityManager.
This happens to work as unwrap() is able to cast it to _EntityManager_ (the interface), but this only works if the actual implementation happens to be an _EntityManagerImpl_. So.. shouldn't you rather cast explicitly?

**Comment by Sanne** [comment]
I'd like to merge something like this, but it feels like a temporary solution as we can't really guarantee that these method calls stay in synch, and it's hard to test.
For the long term should we also pursue having a nice reusable decorator in ORM codebase? that would also:
 - allow us to stay away from the internal implementation classes
 - make ORM's code more readable in this regard

The one problem I see, is that if we improve on this in ORM's code, we'll need to delegate the unwrap() call too, and the trick you do in the above comment would break.

**Comment by Sanne** [comment]
BTW I don't see my comment above as a blocker: just wondering how to improve in next.gen ORM.

If you think the _unwrap_ catch is a good compromise for the time being, I'll merge it.

File: core/src/main/java/org/hibernate/ogm/jpa/impl/OgmEntityManagerFactory.java (hunk 1 above)
Comment (hferentschik) [inline core/src/main/java/org/hibernate/ogm/jpa/impl/OgmEntityManagerFactory.java]: Unwrapping is what one should do, but I agree the unwrap method is fishy. Unwrapping should check against `EntityManagetImpl` explicitly. In fact it should probably only be possible to unwrap to `org.hibernate.jpa.spi.AbstractEntityManagerImpl

**Comment by hferentschik** [comment]
> I'd like to merge something like this,

do it then ;-)

> but it feels like a temporary solution as we can't really guarantee that these method calls stay in synch, and it's hard to test.

Sure, there is no guarantee that it won't stay in sync, but short of probably quite extensive surgery in the ORM/OGM code base I am not sure how you would solve that. I am not even sure how a "delegate" in ORM would help. 

> If you think the unwrap catch is a good compromise for the time being,

Not sure what you mean, but I actually removed the unwrap and use casts now.

**Comment by hferentschik** [comment]
Pushed an updated version using casts instead of `unwrap`.

**Comment by Sanne** [comment]
thanks, merged!

You think we should enhance the unwrap implementation to "extract" implementation types?

**Comment by hferentschik** [comment]
> You think we should enhance the unwrap implementation to "extract" implementation types?

Yes, but that's a ORM issue right?

**Comment by Sanne** [comment]
Sure, just wondering if you think we should follow up on that. I'm not sure about the intentions of the _unwrap_ method. Personally I think I like it that you can only extract an interface, but you seem to have a different opinion? Open an ORM issue if you can make a good argument for it, since 5 is coming soon..

**Comment by hferentschik** [comment]
> Open an ORM issue if you can make a good argument for it, since 5 is coming soon..

FYI - HHH-9665

**Pull request closed without merging**